File: services/payment-orchestrator-java/src/main/java/com/acme/payments/orchestrator/infrastructure/CorrelationIdFilter.java
package com.acme.payments.orchestrator.infrastructure;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.slf4j.MDC;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;
import java.util.UUID;

@Component
public class CorrelationIdFilter extends OncePerRequestFilter {
    public static final String HEADER = "X-Correlation-Id";
    private static final String MDC_KEY = "correlationId";
    private static final Logger LOGGER = LoggerFactory.getLogger(CorrelationIdFilter.class);

    @Override
    protected void doFilterInternal(
            HttpServletRequest request,
            HttpServletResponse response,
            FilterChain filterChain
    ) throws ServletException, IOException {
        long startedAt = System.nanoTime();
        String correlationId = request.getHeader(HEADER);
        if (!isSafeCorrelationId(correlationId)) {
            correlationId = UUID.randomUUID().toString();
        }

        MDC.put(MDC_KEY, correlationId);
        response.setHeader(HEADER, correlationId);
        try {
            filterChain.doFilter(request, response);
        } finally {
            LOGGER.info("http_outcome service=payment-orchestrator method={} route={} status={} duration_ms={} correlation_id={}",
                    request.getMethod(), request.getRequestURI(), response.getStatus(),
                    (System.nanoTime() - startedAt) / 1_000_000, correlationId);
            MDC.remove(MDC_KEY);
        }
    }

    private boolean isSafeCorrelationId(String value) {
        return value != null && value.length() <= 128 && value.matches("[A-Za-z0-9._:-]+");
    }
}

File: services/pix-boleto-kotlin/src/main/kotlin/com/acme/payments/pixboleto/infrastructure/CorrelationIdFilter.kt
package com.acme.payments.pixboleto.infrastructure

import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.LoggerFactory
import org.slf4j.MDC
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import java.util.UUID

@Component
class CorrelationIdFilter : OncePerRequestFilter() {
    private val logger = LoggerFactory.getLogger(javaClass)

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain,
    ) {
        val startedAt = System.nanoTime()
        val correlationId = request.getHeader(HEADER)?.takeIf(::isSafeCorrelationId)
            ?: UUID.randomUUID().toString()

        MDC.put(MDC_KEY, correlationId)
        response.setHeader(HEADER, correlationId)
        try {
            filterChain.doFilter(request, response)
        } finally {
            logger.info("http_outcome service=pix-boleto method={} route={} status={} duration_ms={} correlation_id={}", request.method, request.requestURI, response.status, (System.nanoTime() - startedAt) / 1_000_000, correlationId)
            MDC.remove(MDC_KEY)
        }
    }

    private fun isSafeCorrelationId(value: String?): Boolean =
        value != null && value.length <= 128 && value.matches(Regex("[A-Za-z0-9._:-]+"))

    companion object {
        const val HEADER = "X-Correlation-Id"
        private const val MDC_KEY = "correlationId"
    }
}
